source/common/api/posix/os_sys_calls_impl.cc

```diff
@@ -42,12 +42,24 @@ SysCallIntResult OsSysCallsImpl::close(os_fd_t fd) {
 }
 
 SysCallSizeResult OsSysCallsImpl::writev(os_fd_t fd, const iovec* iov, int num_iov) {
-  const ssize_t rc = ::writev(fd, iov, num_iov);
+  ssize_t rc;
+  if (num_iov == 1) {
+    // Avoid paying the VFS overhead when there is only one IO buffer to work with
+    rc = ::write(fd, iov[0].iov_base, iov[0].iov_len);
```

Comment:
`OsSysCallsImpl` is just simple wrap of syscall, we shouldn't hack it here. It seems like the only place write one slice with `writev` is SSL socket transport socket.

https://github.com/envoyproxy/envoy/blob/0086411dd85231434c28ffb6af0ebeaf49e6b02f/source/extensions/transport_sockets/tls/io_handle_bio.cc#L74

if this is the right place, we should change the io handle bio to use `write`.

Maybe we need some benchmark on generic TLS, which probably needs maintainers input.

Reply:
> if this is the right place, we should change the io handle bio to use `IoHandle::write`.
> 

And that needs an different read/write interface IoHandle for accept plain buffer.

Reply:
Thanks for the input!

Actually after talking with my colleague, it looks like that the perf improvement doesn't come from replacing `writev()` with `write()` (the scatter-gather doesn't introduce much overhead), but from replacing `writev()` with `send()`, which will reduce some kernel VFS handling overhead.

I'm thinking whether to move the logic to:
https://github.com/envoyproxy/envoy/blob/0086411dd85231434c28ffb6af0ebeaf49e6b02f/source/common/network/io_socket_handle_impl.cc#L118